desktopexporter/internal/store/store.go

```diff
@@ -1,134 +1,288 @@
-			TraceID: spanData.TraceID,
-			Spans:   []telemetry.SpanData{},
+	con, err := s.connector.Connect(ctx)
+	if err != nil {
+		log.Fatalf("could not connect to database: %s", err.Error())
+	}
+	defer con.Close()
+
+	appender, err := duckdb.NewAppenderFromConn(con, "", "spans")
+	if err != nil {
+		log.Fatalf("could not create new appender for spans: %s", err.Error())
+	}
+	defer appender.Close()
+
+	for _, span := range spans {
+		attributes, err := json.Marshal(span.Attributes)
+		if err != nil {
+			log.Fatalf("could not marshal span attributes: %s", err.Error())
+		}
+
+		events, err := json.Marshal(span.Events)
+		if err != nil {
+			log.Fatalf("could not marshal span events: %s", err.Error())
+		}
+
+		links, err := json.Marshal(span.Links)
+		if err != nil {
+			log.Fatalf("could not marshal span links: %s", err.Error())
+		}
+
+		resourceAttributes, err := json.Marshal(span.Resource.Attributes)
+		if err != nil {
+			log.Fatalf("could not marshal resource attributes: %s", err.Error())
+		}
+
+		scopeAttributes, err := json.Marshal(span.Scope.Attributes)
+		if err != nil {
+			log.Fatalf("could not marshal scope attributes: %s", err.Error())
+		}
+
+		if err := appender.AppendRow(
+			span.TraceID,
+			span.TraceState,
+			span.SpanID,
+			span.ParentSpanID,
+			span.Name,
+			span.Kind,
+			span.StartTime,
+			span.EndTime,
+			string(attributes),
+			string(events),
+			string(links),
+			string(resourceAttributes),
+			span.Resource.DroppedAttributesCount,
+			span.Scope.Name,
+			span.Scope.Version,
+			string(scopeAttributes),
+			span.Scope.DroppedAttributesCount,
+			span.DroppedAttributesCount,
+			span.DroppedEventsCount,
+			span.DroppedLinksCount,
+			span.StatusCode,
+			span.StatusMessage,
+		); err != nil {
+			log.Fatalf("could not append row to spans: %s", err.Error())
 		}
 	}
-	traceData.Spans = append(traceData.Spans, spanData)
-	store.TraceMap[spanData.TraceID] = traceData
 }
 
-func (store *Store) GetTrace(traceID string) (telemetry.TraceData, error) {
-	store.mut.Lock()
-	defer store.mut.Unlock()
+func (s *Store) GetTrace(ctx context.Context, traceID string) (telemetry.TraceData, error) {
+	s.mut.Lock()
+	defer s.mut.Unlock()
 
-	trace, traceExists := store.TraceMap[traceID]
-	if !traceExists {
-		return telemetry.TraceData{}, telemetry.ErrTraceIDNotFound
+	conn, err := s.db.Conn(ctx)
+	if err != nil {
+		log.Fatalf("could not connect to database: %s", err.Error())
 	}
+	defer conn.Close()
 
-	return trace, nil
-}
+	trace := telemetry.TraceData{
+		TraceID: traceID,
+		Spans:   []telemetry.SpanData{},
+	}
 
-func (store *Store) GetRecentTraces(traceCount int) []telemetry.TraceData {
-	store.mut.Lock()
-	defer store.mut.Unlock()
+	rows, err := conn.QueryContext(ctx, SELECT_TRACE, traceID)
+	if err == sql.ErrNoRows {
+		return trace, telemetry.ErrTraceIDNotFound
+	} else if err != nil {
+		log.Fatalf("could not retrieve spans: %s", err.Error())
+	}
+
+	for rows.Next() {
+		span := telemetry.SpanData{}
```

Comment:
Wow, this is kind of painful, but looks right. Factoring this out into a helper function to get a span from a row result might help the code read easier.